signals/signal_exporter.py

```diff
@@ -0,0 +1,383 @@
+        input of expected hostnames (subscribers) as well as a tag.
+        """
+        if not self._valid_str_list(legal_events):
+            print("ERROR: 'legal_events' arg must be a list of string event names")
+            return
+
+        if expected_hosts:
+            if not self._valid_str_list(expected_hosts):
+                print("ERROR: 'expected_hosts' arg must be a list of string hostnames")
+                return
+            for host in expected_hosts:
+                self.subs.append(host + "-resp")
+
+        self.legal_events = legal_events
+        sig = self._sig_builder(event="initialization", tag=tag)
+        self._fetch_responders()
+        self.redis.publish(channel="event-signal-pubsub", message=sig.to_json_str())
+
+    def shutdown(self, tag: str = None) -> None:
+        """
+        Shuts down initialization response listener (stops accepting subscribers).
+        Wipes the subscriber list and publishes a shutdown message.
+        """
+        sig = self._sig_builder(event="shutdown", tag=tag)
+        self.init_listener.stop()
+        self.subs = []
+        self.redis.publish(channel="event-signal-pubsub", message=sig.to_json_str())
+
+
+class SignalResponder:
+    """
+    A signal management object for tools that wish to respond to event/state signals.
+    Can be used both for listening for signals as well as responding to them. Also
+    allows for locking onto specific tags/publisher_ids. Uses the standard signal
+    response protocol for all published messages.
+    """
+
+    def __init__(self, redis_host: str = "localhost", redis_port: int = 6379) -> None:
+        """
+        Sets exporter object fields and generates unique tool_id.
+        Allows for specification of redis host/port.
+        """
+        self.redis = redis.Redis(host=redis_host, port=redis_port, db=0)
+        self.subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+        self.subscriber.subscribe("event-signal-pubsub")
+        self.tool_id = platform.node() + "-resp"  # ADD UUID HERE?
+        self.locked_id = None
+        self.locked_tag = None
+
+    def _parse_signal(self, signal: Dict) -> Dict:
+        """
+        Validates received signal. Returns payload if valid.
+        Also applies tag/publisher_id filters if added.
+        """
+        try:
+            data = json.loads(signal["data"])
+        except ValueError:
+            return None
+        # FIXME - Replace below line, maybe with dataclasses.fields()?
+        check_set = set(
+            [
+                "publisher_id",
+                "process_name",
+                "event",
+                "runner_host",
+                "sample_no",
+                "tag",
+                "metadata",
+            ]
+        )
+        if set(data.keys()) == check_set or check_set - set(data.keys()) == {
+            "metadata"
+        }:
+            if (not self.locked_id or self.locked_id == data["publisher_id"]) and (
```

Comment:
I totally understand what you are trying to do here, but can we revisit this? From my understanding here is the goal of the function:
* Validate that the signal is JSON
* Validate that the signal has all of the needed keys for constructing a signal.
* Filter out signals that aren't part of the ``locked_id`` or ``locked_tag``

What I'm confused about is:
* Why do we ignore data that are valid JSON, yet aren't valid signals, by returning ``None`` after the first if statement?
* Why don't we create a ``Signal`` instance from the given data and instead keep it as a ``dict``?
* The listen method is going to yield all of these ``None`` values when it gets signals to ignore, right? At least I think that's how it'll work.

So here is what I think would make this functionality really easy to extend and to read:

* Let ``_parse_signal`` be just a method that returns a ``Signal`` instance from the given data ``Dict``, raising ``ValueError`` if something is wrong.
* Create another method called ``_is_ignored`` or ``_is_target`` or something, which does the filtering by returning ``True/False`` on whether there is a match for ``locked_tag`` or ``locked_id`` when given a ``Signal`` instance
* Change ``listen`` to only yield when your ignore method says that the signal is a part of the locked tag and id.

This separates parsing and ignoring functionality into these bite-size pieces.

Reply:
When I created the signal and response dataclasses, I had initially thought of them as internal objects, though you raise a fair point. Do you think it would be better for the user to receive a Signal instead of a dict with the fields? If so I can make that change, and in the meantime I'll also separate out the componentsof this process

Reply:
Fixed everything else but returning a signal (an easy change though). Curious to see if you think returning a Signal would be more convenient or confusing for a user.

Reply:
Is the user intended to be only other Python programs, or any application that can consume a JSON?

Reply:
Awesome, thank you for taking the time to do that change. I think @mfleader hit it right on the money with this question. I think that it would be safe to assume that any user using the Python library would be expected to use the Python dataclasses rather than a JSON dict, whereas we can leave the JSON dict usage to other non-python users of the protocol. Additionally, I like the more OO approach of using the signal class rather than the dict. Makes the returned data easy to document and read from a user perspective IMHO.